routing/payment_lifecycle.go

```diff
@@ -2,346 +2,592 @@ package routing
+		// either we've gone past the payment attempt timeout, or the
+		// router is exiting. In either case, we'll stop this payment
+		// attempt short. If a timeout is not applicable, timeoutChan
+		// will be nil.
 		select {
-		case result, ok = <-resultChan:
-			if !ok {
-				return [32]byte{}, nil, htlcswitch.ErrSwitchExiting
+		case <-p.timeoutChan:
+			errStr := fmt.Sprintf("payment attempt not completed " +
+				"before timeout")
+			log.Warnf(errStr)
+
+			saveErr := p.router.cfg.Control.Fail(
+				p.paymentHash, channeldb.FailureReasonTimeout,
+			)
+			if saveErr != nil {
+				return [32]byte{}, nil, saveErr
 			}
 
+			continue
+
+		// The payment will be resumed from the current state after
+		// restart.
 		case <-p.router.quit:
 			return [32]byte{}, nil, ErrRouterShuttingDown
+
+		// Fall through if we haven't hit our time limit.
+		default:
 		}
 
-		// In case of a payment failure, we use the error to decide
-		// whether we should retry.
-		if result.Error != nil {
-			log.Errorf("Attempt to send payment %x failed: %v",
-				p.payment.PaymentHash, result.Error)
+		// Create a new payment attempt from the given payment session.
+		rt, err := p.paySession.RequestRoute(
+			remValue, feeLimit, uint32(p.currentHeight),
+		)
+		if err != nil {
+			log.Warnf("Failed to find route for payment %x: %v",
+				p.paymentHash, err)
+
+			// There is no route to try, and we have no active
+			// shards. This means that there is no way for us to
+			// send the payment, so mark it failed with no route.
+			if !activeShards {
+				failureCode := errorToPaymentFailure(err)
+				log.Debugf("Marking payment %v permanently "+
+					"failed with no route: %v",
+					p.paymentHash, failureCode)
+
+				saveErr := p.router.cfg.Control.Fail(
+					p.paymentHash, failureCode,
+				)
+				if saveErr != nil {
+					return [32]byte{}, nil, saveErr
+				}
+
+				continue
+			}
 
-			err = p.failAttempt(result.Error)
-			if err != nil {
-				return [32]byte{}, nil, err
+			// We still have active shards, we'll wait for an
+			// outcome to be available before retrying.
+			select {
+			case err := <-shardErrors:
+				if err != nil {
+					return [32]byte{}, nil, err
+				}
+
+			case <-p.router.quit:
+				return [32]byte{}, nil, ErrRouterShuttingDown
 			}
 
+			continue
+		}
+
+		// We found a route to try, launch a new shard.
+		attempt, outcome, err := shardHandler.launchShard(rt)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		// We ew encountered a non-critical error when launching the
+		// shard, handle it
+		if outcome.err != nil {
+			log.Warnf("Failed to launch shard %v for "+
+				"payment %v: %v", attempt.AttemptID,
+				p.paymentHash, outcome.err)
+
 			// We must inspect the error to know whether it was
 			// critical or not, to decide whether we should
 			// continue trying.
-			if err := p.handleSendError(result.Error); err != nil {
+			err := shardHandler.handleSendError(
```

Comment:
The reason that `handleSendError` is a separate call and not integrated with `launchShard`, is because of SendToRoute?

Reply:
I got a bit confused because `collectResultAsync` is calling `handleSendError` too. So `handleSendError` is called both internally and externally. Maybe a wrapper around `launchShard` and `handleSendError` would be nice, to keep the calls to `handleSendError` internal as much as possible